docs/design_proposals/log-selectors.md

```diff
@@ -0,0 +1,128 @@
+```yaml
+log:
+  # a list of deployment.spec.selector items, rules are OR'ed
+  - matchExpressions: # in one item, rules are AND'ed
+      - key: XY
+        values: ['val1', 'val2']
+        operator: In NotIn Exists DoesNotExist
+    # or
+    matchLabels: # in one item, rules are AND'ed
+      tail: "true"
+```
+This structure is based on the [standard selector spec](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors) for k8s resources.
+Using a standard k8s configuration structure increases the usability for our users.
+For one selector spec item, the semantics should be exactly the same as for deployments, i.e. several conditions are AND'ed. Quoting the kubernetes docs (`kubectl explain deployment.spec.selector`):
+```txt
+KIND:     Deployment
+VERSION:  extensions/v1beta1
+
+RESOURCE: selector <Object>
+
+DESCRIPTION:
+     Label selector for pods. Existing ReplicaSets whose pods are selected by
+     this will be the ones affected by this deployment.
+
+     A label selector is a label query over a set of resources. The result of
+     matchLabels and matchExpressions are ANDed. An empty label selector matches
+     all objects. A null label selector matches no objects.
+
+FIELDS:
+   matchExpressions     <[]Object>
+     matchExpressions is a list of label selector requirements. The requirements
+     are ANDed.
+
+   matchLabels  <map[string]string>
+     matchLabels is a map of {key,value} pairs. A single {key,value} in the
+     matchLabels map is equivalent to an element of matchExpressions, whose key
+     field is "key", the operator is "In", and the values array contains only
+     "value". The requirements are ANDed
+```
+
+In addition, Skaffold needs to support the use-case to add further pods to a selection.
+To address that, the log spec has to be a list of selector specs which is OR'ed.
+For example, to add the traefik controller to the watch list:
+```yaml
+log:
+  - matchExpressions:
+      - key: app
+        values: ['traefik']
+        operator: In
+  - matchLabels:
+      tail: "true"
+```
+
+This config change should be made backwards compatible by defaulting the log selector to
+```yaml
+log:
+  - matchLabels:
+      tail: "true"
+```
+Skaffold already labels all deployed pods with `tail=true` unless run with the `--tail=false` CLI option.
+
+### Open Issues/Question
+
+Please list any open questions here in the format.
+
+**\<Is the `tail=true` label good enough?\>** In principle, it would be possible that other pods also have a `tail=true` label by accident (for example a previous `skaffold run --tail` followed by `skaffold dev`).
+Do we need to exclude such cases, e.g. by adding a unique `skaffold-run: <uuid>` label?
+
+Resolution: A UUID label is not required for now but can be added later when users request it.
+
+## Implementation plan
+1. Switch the selection of pods for the log selector from image lists to `tail=true` label (#1910).
```

Comment:
There's an issue with `helm`: labels are applied to resources after they are created. not before. This means that we won't see the logs during a short period of time.

Reply:
@dgageot 
> There's an issue with helm : labels are applied to resources after they are created. not before.

You mean this is a restriction of `helm`  itself or how skaffold handles `helm`  deployments?

Reply:
Because `helm` templates are resolved cluster side, skaffold can't apply the labels on the yaml before sending to k8s, the way it does with `kubectl` and `kustomize`.

Reply:
I think this is a minor issue. Once the label is applied by `helm`, the logs will be tailed starting from the global startTime (pkg/skaffold/kubernetes/log.go:50). I just tried this with the following patch based on PR #1910
```patch
diff --git a/examples/helm-deployment/Dockerfile b/examples/helm-deployment/Dockerfile
index c1c9be3e..2307eff6 100644
--- a/examples/helm-deployment/Dockerfile
+++ b/examples/helm-deployment/Dockerfile
@@ -1,3 +1,7 @@
-FROM nginx:stable
+FROM golang:1.10.1-alpine3.7 as builder
+COPY main.go .
+RUN go build -o /app main.go
 
-COPY static /usr/share/nginx/html/
+FROM alpine:3.7
+CMD ["./app"]
+COPY --from=builder /app .
diff --git a/examples/helm-deployment/main.go b/examples/helm-deployment/main.go
new file mode 100644
index 00000000..46d5cbd9
--- /dev/null
+++ b/examples/helm-deployment/main.go
@@ -0,0 +1,14 @@
+package main
+
+import (
+	"fmt"
+	"time"
+)
+
+func main() {
+	for i := 0; ; i++ {
+		fmt.Printf("Hello world! %d\n", i)
+
+		time.Sleep(time.Second * 1)
+	}
+}
diff --git a/examples/helm-deployment/skaffold-helm/templates/deployment.yaml b/examples/helm-deployment/skaffold-helm/templates/deployment.yaml
index ad989438..f1b4442b 100644
--- a/examples/helm-deployment/skaffold-helm/templates/deployment.yaml
+++ b/examples/helm-deployment/skaffold-helm/templates/deployment.yaml
@@ -14,6 +14,7 @@ spec:
       labels:
         app: {{ template "skaffold-helm.name" . }}
         release: {{ .Release.Name }}
+        tail: "true"
     spec:
       containers:
         - name: {{ .Chart.Name }}
```
All logs starting from iteration `0` are shown when using the label selector approach.
Therefore, there will be a small delay until the logs are shown, but nothing will be lost.

**What bothers me** though, is that `helm` deployments don't seem to apply the `tail=true` label to deployed pods. I had to manually apply that in the above patch. Is that intentional?

Reply:
So this is a real problem, see #2074 for further information. I also opened PR #2075 which would fix this, so that the label selection with helm deployments works again.